File: SD2-viola_jones/testes/test_yunet.py
import cv2 as cv
import numpy as np

# Carregar o modelo YuNet
model = 'face_detection_yunet_2023mar.onnx'
input_size = (640, 640)
face_detector = cv.FaceDetectorYN.create(
    model, "", input_size, score_threshold=0.8, nms_threshold=0.3,
    top_k=5000, backend_id=cv.dnn.DNN_BACKEND_OPENCV, target_id=cv.dnn.DNN_TARGET_CPU
)



# Captura de vídeo da webcam
cap = cv.VideoCapture(0)

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        print("Falha na captura de vídeo")
        break

    # Pré-processamento (opcional)
    # frame = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
    # frame = cv.equalizeHist(frame)
    # frame = cv.cvtColor(frame, cv.COLOR_GRAY2BGR)

    # Redimensionar o frame para o tamanho de entrada
    resized_frame = cv.resize(frame, input_size)

    
    faces = face_detector.detect(resized_frame)

    if faces[1] is not None:
        for face in faces[1]:
            # Escalar as coordenadas de volta para o tamanho original
            face = face.astype(int)
            x, y, w, h = face[:4]
            x = int(x * frame.shape[1] / input_size[0])
            y = int(y * frame.shape[0] / input_size[1])
            w = int(w * frame.shape[1] / input_size[0])
            h = int(h * frame.shape[0] / input_size[1])

            # Desenhar a caixa delimitadora ao redor do rosto
            cv.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)

            # Desenhar os pontos dos olhos, nariz e boca
            landmarks = face[4:14].reshape((5, 2))
            for (lx, ly) in landmarks:
                lx = int(lx * frame.shape[1] / input_size[0])
                ly = int(ly * frame.shape[0] / input_size[1])
                cv.circle(frame, (lx, ly), 2, (0, 0, 255), -1)

           

    # Exibir o vídeo com as detecções
    cv.imshow('Detecção Facial com YuNet', frame)

    if cv.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv.destroyAllWindows()

File: SD2-viola_jones/raspb_3/cpp/test_yunet.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/objdetect.hpp>
#include <iostream>

int main() {
    // Carregar o modelo de detecção facial
    std::string model = "face_detection_yunet_2023mar.onnx";
    cv::Size input_size(320, 320);
    double score_threshold = 0.9;
    double nms_threshold = 0.3;
    int top_k = 5000;
    int backend_id = cv::dnn::DNN_BACKEND_OPENCV;
    int target_id = cv::dnn::DNN_TARGET_CPU;

    cv::Ptr<cv::FaceDetectorYN> face_detector = cv::FaceDetectorYN::create(
        model, "", input_size, score_threshold, nms_threshold, top_k, backend_id, target_id);

    // Captura de vídeo da webcam
    cv::VideoCapture cap(0);
    if (!cap.isOpened()) {
        std::cout << "Não foi possível conectar à webcam" << std::endl;
        return -1;
    }

    // Obter as dimensões do frame
    cv::Mat frame;
    cap >> frame;
    if (frame.empty()) {
        std::cout << "Falha na captura de vídeo" << std::endl;
        return -1;
    }

    int frame_width = frame.cols;
    int frame_height = frame.rows;
    face_detector->setInputSize(cv::Size(frame_width, frame_height));

    while (true) {
        cap >> frame;
        if (frame.empty()) {
            std::cout << "Falha na captura de vídeo" << std::endl;
            break;
        }

        cv::Mat faces;
        face_detector->detect(frame, faces);

        if (!faces.empty()) {
            for (int i = 0; i < faces.rows; ++i) {
                // Extrair as coordenadas da caixa delimitadora
                float x = faces.at<float>(i, 0);
                float y = faces.at<float>(i, 1);
                float w = faces.at<float>(i, 2);
                float h = faces.at<float>(i, 3);
                cv::Rect face_rect(cv::Point(cvRound(x), cvRound(y)), cv::Size(cvRound(w), cvRound(h)));
                cv::rectangle(frame, face_rect, cv::Scalar(0, 255, 0), 2);

                // Desenhar os pontos de referência (landmarks)
                for (int k = 4; k < 14; k += 2) {
                    float lx = faces.at<float>(i, k);
                    float ly = faces.at<float>(i, k + 1);
                    cv::circle(frame, cv::Point(cvRound(lx), cvRound(ly)), 2, cv::Scalar(0, 0, 255), -1);
                }
            }
        }

        cv::imshow("Detecção Facial com YuNet", frame);

        if (cv::waitKey(1) == 'q') {
            break;
        }
    }

    cap.release();
    cv::destroyAllWindows();
    return 0;
}
